content/posts/2024-12-04-improve-clang-doc.md

```diff
@@ -0,0 +1,125 @@
+
+For example, if we had the following project: 
+
+
+```cpp
+//File: Base.cpp
+
+class Base {}
+```
+
+
+```cpp
+//File: A.cpp
+
+#include "Base.cpp"
+
+...
+
+```
+
+```cpp
+//File: B.cpp
+
+#include "Base.cpp"
+
+...
+```
+
+In this case, the ASTVisitor invoked by Clang-Doc would visit the serialized Base class three times, once when it is parsing Base.cpp, another when its visiting A.cpp then B.cpp. The problem was that there was no mechanism to identify declarations that we had already seen. The optimization ended up being a simple memoization dictionary which kept track of a list of declaration that Clang-Doc had visited. 
+
+
+
+Here is a plot of the benchmarking numbers:
+
+<div style="margin:0 auto;">
+  <img src="/img/Clang-Doc-benchmark-numbers.png"><br/>
+</div>
+
+
+### Added Template Mustache HTML Backend
+
+
+Clang-Doc originally used an ad-hoc method of generating HTML. I introduced a templating language as a way of reducing project complexity and reducing the ease of development. Two RFCs were made before arriving at the idea of introducing Mustache as a library. Originally the idea was to introduce a custom templating language, however, upon further discussion, it was decided that the complexity of designing and implementing a new templating language was too much.
+An LLVM community member suggested using Mustache as a templating language. 
+Mustache was the ideal choice since it was very simple to implement, and has a well defined spec that fit what was needed for Clang-Doc’s use case. The feedback on the [RFC](https://discourse.llvm.org/t/rfc-add-template-mustache-language-to-the-support-library/82439/18) was generally positive. While there was some resistance regarding the inclusion of an HTML support library in LLVM, this concern stemmed partly from a lack of awareness that HTML generation already occurs in several parts of LLVM. Additionally, the introduction of Mustache has the potential to simplify other HTML-related use cases.
+In terms of engineering wins, this library was able to cut the direct down on the HTML backend significantly dropping 500 lines of code compared to the original Clang-Doc HTML backend. This library was also designed for general-purpose use around LLVM since there are numerous places in LLVM where various tools generate HTML in its way. Using the Mustache templating library would be a nice way to standardize the codebase. 
+
+### Improve Clang-Doc HTML Output
+
+The previous version of Clang-Doc’s output was a pretty minimal bare bones implementation. It had a sidebar that contained every single declaration within the project which created a large unnavigable UI. Typedef documentation was missing, plus method documentation was missing details such as whether or not the method was a const or virtual. There was no linking between other declarations in the project and there was no syntax highlighting on any language construct.
```

Comment:
```suggestion
The previous version of Clang-Doc’s HTML output was a pretty minimal, bare bones implementation. It had a sidebar that contained every single declaration within the project which created a large unnavigable UI. Typedef documentation was missing, plus method documentation was missing details such as whether or not the method was a const or virtual. There was no linking between other declarations in the project and there was no syntax highlighting on any language construct.
```

Up to you on this, but I'd suggest discussing the UI more since you feel strongly about it. For instance you mention that the UI was hard to use, but I feel like you can lean in here and say more about the what/why. This is a personal style choice though, so it's your call.